### PR TITLE
EWB-3763 Protection extensions

### PR DESCRIPTION
**Pull request by vincewhite**
# Description

The python implementation of EWB-3757 (Protection model change) https://github.com/zepben/evolve-sdk-jvm/pull/158

# Associated tasks

The jvm-sdk implementation: https://github.com/zepben/evolve-sdk-jvm/pull/158

Requires: https://github.com/zepben/evolve-grpc/pull/84 v0.27.0?

# Checklist

If any of these are not applicable, strikethrough the line `~like this~`. **Do not delete it!**. Let the reviewer decide if you should have done it.

### Code
- [~] I have performed a self review of my own code (including checking issues raised when creating the PR).
- [x] I have added/updated unit tests for these changes, and if not I have explained why they are not necessary.
- [x] I have commented my code in any hard-to-understand or hacky areas.
- [x] I have handled all new warnings generated by the compiler or IDE.
- [x] I have rebased onto the target branch (usually main).
      
### Documentation
- [~] I have updated the changelog.
- [x] I have updated any documentation required for these changes.

# Breaking Changes
- [x] I have considered if this is a breaking change and will communicate it with other team members if so.

* Removed `ProtectionEquipment`.
* Change of inheritance: `CurrentRelay` &rarr; `ProtectionEquipment`.
  becomes `CurrentRelay` &rarr; `ProtectionRelayFunction`.
* Removed symmetric relation `ProtectionEquipment` &harr; `ProtectedSwitch`.
* Renamed `CurrentRelayInfo` to `RelayInfo`.
  * The override `CurrentRelay.relay_info` has been moved from `CurrentRelay` to its new parent class, `ProtectionRelayFunction`.
  * Renamed `RelayInfo.remove_delay` to `RelayInfo.remove_delay_at`. The original method name has been repurposed to remove a delay by its value rather than its
    index.
* Reworked values for enumerable type `ProtectionKind`.

# Other Things I Touched
Since `test.services.common.service_comparator_validator.ServiceComparatorValidator` now has the ability to handle expected side effects when validating the properties of objects via the `expected_differences` parameter. 

I updated the `NetworkServiceComparator` to include the properties that "override" `asset_info`  e.g. `ProtectionRelayFunction.relay_info` or `PowerTransformerInfo.power_transformer_info`. 

This also included updating `ServiceComparatorValidator` to test/validate comparing these properties and their expected side effects on `asset_info`(and vice versa).

### CODE REVIEW

**Comment by ryanjcoleman** [comment]
Task linked: [EWB-3763 Update Python SDK](https://app.clickup.com/t/6929263/EWB-3763)

**Comment by ryanjcoleman** [comment]
Task linked: [EWB-3757 Update JVM SDK](https://app.clickup.com/t/6929263/EWB-3757)

File: src/zepben/evolve/model/cim/iec61970/base/wires/power_transformer.py [hunk 1]
```diff
@@ -211,6 +211,26 @@ class TransformerEnd(IdentifiedObject):
...
+        """
+        The terminal of the transformer that this end is associated with
+        """
+        return self._terminal
+
+    @terminal.setter
+    def terminal(self, value: Optional[Terminal]):
+        if value is not None:
+            require(value.conducting_equipment is None or isinstance(value.conducting_equipment, PowerTransformer),
+                    lambda: f"Cannot assign {self.__class__.__name__}[{self.mrid}] to {value.__class__.__name__}[{value.mrid}], which is connected to a " +
```
[the remark this answers is not in the document]
Comment (Lax125) [reply src/zepben/evolve/model/cim/iec61970/base/wires/power_transformer.py:230]: yep

File: src/zepben/evolve/model/cim/iec61970/base/protection/protection_relay_function.py [hunk 2]
```diff
@@ -0,0 +1,423 @@
...
+    @property
+    def thresholds(self) -> Generator[RelaySetting, None, None]:
+        """
+        Yields all the thresholds[:class:`RelaySettings<RelaySetting>`] for this :class:`ProtectionRelayFunction`. The order of thresholds corresponds to the order of time limits.
+
+        :return: A generator that iterates over all thresholds[:class:`RelaySettings<RelaySetting>`] for this relay function.
+        """
+        return ngen(self._thresholds)
+
+    def add_threshold(self, threshold: RelaySetting, sequence_number: int = None) -> ProtectionRelayFunction:
```
[the remark this answers is not in the document]
Comment (Lax125) [reply src/zepben/evolve/model/cim/iec61970/base/protection/protection_relay_function.py:113]: They mean the same thing, but for some reason terminals are 1-indexed. Not sure how it ended up that way, I just copied how it works with locations and position points.

File: src/zepben/evolve/model/cim/iec61970/base/protection/protection_relay_function.py [hunk 3]
```diff
@@ -0,0 +1,423 @@
...
+        """
+        Add a threshold[:class:`RelaySetting`] to this :class:`ProtectionRelayFunction`'s list of thresholds.
+
+        :param threshold: The threshold[:class:`RelaySetting`] to add to this :class:`ProtectionRelayFunction`.
+        :param sequence_number: The sequence number of the `threshold` being added.
+        :return: A reference to this :class:`ProtectionRelayFunction` for fluent use.
+        """
+        if sequence_number is None:
+            sequence_number = self.num_thresholds()
+        require(0 <= sequence_number <= self.num_thresholds(),
```
[the remark this answers is not in the document]
Comment (Lax125) [reply src/zepben/evolve/model/cim/iec61970/base/protection/protection_relay_function.py:123]: I don't see a use case for that. The order only matters for matching up with the time limits (e.g. 5A for 2 seconds, 10A for 0.1 seconds)

File: src/zepben/evolve/database/sqlite/readers/network_cim_reader.py [hunk 4]
```diff
@@ -565,23 +572,92 @@ def _load_measurement(self, measurement: Measurement, table: TableMeasurements,
...
+
+    def load_protection_relay_function_thresholds(self, table: TableProtectionRelayFunctionThresholds, rs: ResultSet,
+                                                  set_last_mrid: Callable[[str], str]) -> bool:
+        protection_relay_function_mrid = rs.get_string(table.protection_relay_function_mrid.query_index)
+
+        protection_relay_function = self._ensure_get(protection_relay_function_mrid, ProtectionRelayFunction)
+
+        sequence_number = rs.get_int(table.sequence_number.query_index)
+        unit_symbol = UnitSymbol[rs.get_string(table.unit_symbol.query_index)]
+        value = rs.get_double(table.value.query_index)
```
[the remark this answers is not in the document]
Comment (Lax125) [reply src/zepben/evolve/database/sqlite/readers/network_cim_reader.py:623]: Maybe require a non-nan value in `RelaySetting.__init__(self)`? Similar to `PositionPoint.__init__(self)`

File: src/zepben/evolve/model/cim/iec61970/base/protection/protection_relay_function.py [hunk 5]
```diff
@@ -0,0 +1,423 @@
...
+                self.add_sensor(sensor)
+        if protected_switches is not None:
+            for protected_switch in protected_switches:
+                self.add_protected_switch(protected_switch)
+        if schemes is not None:
+            for scheme in schemes:
+                self.add_scheme(scheme)
+
+        if (time_limits is not None) != (thresholds is not None):
+            raise ValueError(f"Error initializing {self.__class__.__name__}[{self.mrid}]. time_limits and thresholds must be initialized together.")
```
[the remark this answers is not in the document]
Comment (Lax125) [reply src/zepben/evolve/model/cim/iec61970/base/protection/protection_relay_function.py:75]: I don't think it should. The user should be able to specify the thresholds first and add time limits later or vice-versa.

File: src/zepben/evolve/model/cim/iec61970/base/protection/protection_relay_function.py [hunk 6]
```diff
@@ -0,0 +1,423 @@
...
+    def num_thresholds(self) -> int:
+        """
+        Get the number of thresholds for this :class:`ProtectionRelayFunction`.
+
+        :return: The number of thresholds for this `ProtectionRelayFunction`.
+        """
+        return nlen(self._thresholds)
+
+    def get_threshold(self, sequence_number: int) -> Optional[RelaySetting]:
+        # TODO: raise index error or return None, see RelayInfo.get_delay()
```
[the remark this answers is not in the document]
Comment (Lax125) [reply src/zepben/evolve/model/cim/iec61970/base/protection/protection_relay_function.py:140]: raise index error

File: src/zepben/evolve/model/cim/iec61970/base/protection/relay_setting.py [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+    value: float
+    """The value of the setting, e.g voltage, current, etc."""
+
+    name: Optional[str] = None
+    """The name of the setting."""
+
+    def __init__(self):
+        require(self.value is not None and not math.isnan(self.value),
+                lambda: f"RelaySetting.value must be a real number. Provided: {self.value}")
+
```
Comment (Lax125) [inline src/zepben/evolve/model/cim/iec61970/base/protection/relay_setting.py:33]: Nit: Final line should be empty (no spaces/tabs even)

**Review by Lax125 — CHANGES_REQUESTED** [review]
Praise: all in all very thorough. Thanks for adding missing stuff for current/voltage transformers and LV feeders.

File: src/zepben/evolve/model/cim/iec61970/base/protection/distance_relay.py [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+    """The forward reactance (in ohms) that determines the maximum distance along the transmission line in the opposite direction of power flow for which the 
+    relay will provide protection."""
+
+    operation_phase_angle1: Optional[float] = None
+    """The phase angle (in degrees) between voltage and current during normal operating conditions for zone 1 relay."""
+
+    operation_phase_angle2: Optional[float] = None
+    """The phase angle (in degrees) between voltage and current during normal operating conditions for zone 2 relay."""
+
+    operation_phase_angle3: Optional[float] = None
```
[the remark this answers is not in the document]
Comment (Lax125) [reply src/zepben/evolve/model/cim/iec61970/base/protection/distance_relay.py]: Could you do the same for the table class as well (`TableDistanceRelays`)?

File: setup.py [hunk 9]
```diff
@@ -13,7 +13,7 @@
 deps = [
     "requests<2.27.0,>=2.26.0",
     "zepben.auth==0.11.0",
-    "zepben.protobuf==0.26.0",
+    "zepben.protobuf==0.27.0b5",
```
[the remark this answers is not in the document]
Comment (Lax125) [reply setup.py:16]: Version checks out

**Review by Lax125 — DISMISSED** [review]

**Review by kgreav — DISMISSED** [review]

**Review by kgreav — APPROVED** [review]

**Pull request merged**